### PR TITLE
[LLVM][rtsan] Add module pass to initialize rtsan

### PR DESCRIPTION
**Pull request by cjappl**
This allows shared libraries instrumented with RTSan to be initialized.

This approach directly mirrors the approach in Tsan, Asan and many of the other sanitizers:

https://github.com/llvm/llvm-project/blob/c74e2232f226b95d1cf73b9835ec1691a2022010/llvm/lib/Transforms/Instrumentation/ThreadSanitizer.cpp#L175

### CODE REVIEW

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll [hunk 1]
```diff
@@ -25,3 +25,7 @@ attributes #0 = { mustprogress noinline sanitize_realtime optnone ssp uwtable(sy
 ; RealtimeSanitizer pass should insert __rtsan_realtime_exit right before function return
 ; CHECK: call{{.*}}@__rtsan_realtime_exit
 ; CHECK-NEXT: ret{{.*}}void
+
+; RealtimeSanitizer pass should insert call to initialize the runtime
```
Comment (fmayer) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll:29]: maybe the test should specify the context this call is in?

Comment (cjappl) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan.ll]: Good idea!

**Review by fmayer — APPROVED** [review]

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -1024,12 +1024,14 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
             FPM.addPass(BoundsCheckingPass());
           });
 
-    if (LangOpts.Sanitize.has(SanitizerKind::Realtime))
+    if (LangOpts.Sanitize.has(SanitizerKind::Realtime)) {
```
Comment (vitalybuka) [inline clang/lib/CodeGen/BackendUtil.cpp:1027]: @cjappl 
@why do we use this location and not `addSanitizers`?  In particular registerOptimizerLastEPCallback?

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 3]
```diff
@@ -1024,12 +1024,14 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
...
       PB.registerScalarOptimizerLateEPCallback(
           [](FunctionPassManager &FPM, OptimizationLevel Level) {
             RealtimeSanitizerOptions Opts;
             FPM.addPass(RealtimeSanitizerPass(Opts));
           });
+      MPM.addPass(ModuleRealtimeSanitizerPass());
```
Comment (vitalybuka) [inline clang/lib/CodeGen/BackendUtil.cpp:1033]: Also we converted all but tsan from function+module to just module pass. Easier to maintain.

Comment (cjappl) [reply clang/lib/CodeGen/BackendUtil.cpp:1033]: Sounds good, I'll work on swapping rtsan over.

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 2 above)
Comment (cjappl) [reply clang/lib/CodeGen/BackendUtil.cpp:1027]: I'll investigate if moving it is possible! Off the top of my head I don't know if there is a reason.

I'll report back with PRs on either/both of these soon.

Comment (vitalybuka) [reply clang/lib/CodeGen/BackendUtil.cpp:1027]: You need to consider performance vs quality.

For other sanitizers it's essentially:
1. If it's early - better precision - in sense we detect issues in C/C++ program which can be optimized out. (can Realtime related issue be optimized out, and do you care about them at all)
2. If it's late - better performance, but we may miss bugs
I believe performance is ~2x difference, so we can't afford that e.g. in Asan.

**Pull request merged**